### ISSUE TITLE
Support for default parameter values in generic

### ISSUE THREAD

**Issue by trane** [travisbrown/circe#65]
This appears to be mostly shapeless related, but @travisbrown asked me to file this for posterity...

``` scala
case class Foo(s: String, i: Int = 10)
val json = """{"s": "Hello"}"""

// this obviously works
decode[Int => Foo](json).map(_(10))

// but what about cases where you have many default parameters of the same type?
case class Bar(i1: Int, i2: Int = 2, i3: Int = 3, i4: Int = 4)
val json2 = """{"i1": 1}"""

// i would like the following behavior...
"""{"i1": 1, "i4": 44}""".as[Bar] // => Bar(1, 2, 3, 44)

// this would work kind of work, but it overrides the default values
decode[(Int, Int, Int) => Bar].map(_(2, 3, 44))

// alternatively, I could do this to ignore the potentially incoming values:
val dec = Decoder.instance[Bar](c =>
  for {
    i1 <- c.downField("i1").as[Int]
  } yield Bar(i1)
)

// or change my datatype completely:
case class Bar2(i1: Int, i2: Option[Int], i3: Option[Int], i4: Option[Int])
val dec2 = Decoder.instance[Bar2](c =>
  for {
    i1 <- c.downField("i1").as[Int]
    i2 <- c.downField("i2").as[Option[Int]]
    i3 <- c.downField("i3").as[Option[Int]]
    i4 <- c.downField("i4").as[Option[Int]]
  } yield Bar2(i1, i2, i3, i4)
)
```

**Comment by alexarchambault** [comment]
Something like https://github.com/milessabin/shapeless/pull/459 would allow to address that (default field values can easily be pulled like in this [test](https://github.com/alexarchambault/shapeless/blob/8ec877d9a6f06dd07a2548f60410eb485b3f28dd/core/src/test/scala/shapeless/derive.scala#L283)).

**Comment by travisbrown** [comment]
One general question: should this be the behavior provided by `generic.auto` (if default values are available in the `Default` instance, use them), or should it be an option provided by `semiauto`?—i.e. something like this:

```scala
implicit val decodeBar: Decoder[Bar] = deriveFor[Bar].decoderWithDefaults
```

**Comment by trane** [comment]
I don't know what the benefit is to having it in `semiauto`. The least surprising option, from my point of view, is to treat `case class X` with/without defaults the same way.

**Comment by travisbrown** [comment]
@trane It's not necessarily obvious that a derived decoder will use default constructor values (when they're needed), and if that's _not_ the behavior you want, it could be an unpleasant surprise. But I agree—I think it makes sense to use them (with plenty of documentation) in `auto`, and to provide something like `decoderWithoutDefaults` in `semiauto`.

**Comment by trane** [comment]
@travisbrown don't trust my point of view, I've been coding in Ruby for the past 2 years -- of course I like magic and surprising behavior!

**Comment by LATaylor-guardian** [comment]
@travisbrown Was this ever added by any chance? I was playing around this morning and hit this issue so was unsure if I am missing something?

**Comment by ilya-murzinov** [comment]
@travisbrown I was figuring out how it can be implemented and ended up with the following.
Inside ```DerivationMacros#decodeHList``` and ```DerivationMacros#decodeCoproduct``` we should check whether the class has default values, and if it has, add them to representation.
Am I on the right path?
If so, can you give any tips in how it can be done?

**Comment by travisbrown** [comment]
This is now available in circe-extras in [0.6.0-RC1](https://github.com/travisbrown/circe/releases/tag/v0.6.0-RC1).
